library/bigip_software.py

```diff
@@ -212,673 +290,1060 @@
+        images = self.client.api.tm.sys.software.images.get_collection()
+        for image in images:
+            if image.version == self.want.version:
+                if image.build == self.want.build:
+                    return True
+        return False
 
-        if force:
-            if hotfix:
-                self.delete(hotfix)
-                total -= 1
-                changed = True
+    def hotfix_exists_on_device(self):
+        hotfixes = self.client.api.tm.sys.software.hotfix_s.get_collection()
+        for image in hotfixes:
+            if image.version == self.want.version:
+                if image.build == self.want.build:
+                    return True
+        return False
 
-            if software:
-                self.delete(software)
-                total -= 1
-                changed = True
+    def load_volume_on_device(self):
+        volume = self.client.api.tm.sys.software.volumes.volume.load(
+            name=self.want.volume)
+        return volume
 
-        if changed:
-            self.wait_for_images(total)
-            changed = False
+    def reboot_volume_on_device(self):
+        self.client.api.tm.sys.software.volumes.exec_cmd(
+            'reboot', volume=self.want.volume
+            )
 
-        if hotfix:
-            if not self.is_hotfix_available(hotfix):
-                self.upload(hotfix)
-                total += 1
-                changed = True
+    def volume_exists_on_device(self):
+        return self.client.api.tm.sys.software.volumes.volume.exists(
+            name=self.want.volume
+        )
 
-        if not self.is_software_available(software):
-            self.upload(software)
-            total += 1
-            changed = True
 
-        if changed:
-            self.wait_for_images(total)
+class RemoteManager(BaseManager):
+    def software_on_volume(self):
+        self.check_product_info()
+        volumes = self.list_volumes_on_device()
+        version = self.want.version
+        build = self.want.build
+        for volume in volumes:
+            if hasattr(volume, 'version') and hasattr(volume, 'build'):
+                if volume.version == version and volume.build == build:
+                    tmp_res = volume.attrs
+                    return Parameters(tmp_res), volume
+
+    def check_product_info(self):
+        hotfix = self.want.hotfix
+        version = self.want.version
+        build = self.want.build
+        state = self.want.state
+
+        if state == 'installed' or state == 'activated':
+            if version is None or build is None:
+                if hotfix:
+                    self.set_product_info(True)
+                else:
+                    self.set_product_info()
+            return False
 
-        status = self.api.System.SoftwareManagement.get_all_software_status()
-        volumes = [x['installation_id']['install_volume'] for x in status if x['active']]
+        if state == 'present' or state == 'absent':
```

Comment:
this state and the above are mutually exclusive, so make this an `elif`

also change the statement to be if state in ['presen...']